### PR TITLE
Diagnostics: Clearer information style

### PR DESCRIPTION
**Pull request by muqiuhan**
Screenshot_20240805_161112

  * Add file content to message.
  * Use more symbols and line breaks to separate messages.

## Description

In this PR, I tried to make some changes to [CompilerDiagnostics.fs](https://github.com/muqiuhan/fsharp/blob/diagnostic-messages-improvements/src/Compiler/Driver/CompilerDiagnostics.fs) to make the compiler print clearer error messages (or warning messages, etc.)

I tried adding a Context field to `FormattedDiagnosticDetailedInfo`, which reads the corresponding content from the target file according to the information range indicated by the Range type, and then adds it to `buf` in `PhasedDiagnostic`.

This is my first time entering the world of the F# compiler, and I know this PR has a lot of issues, but I would like to put forward a small idea about #14832 . ^_^

## Checklist

- [ ] Test cases added
- [ ] Performance benchmarks added in case of performance changes
- [ ] Release notes entry updated

### CODE REVIEW

**Comment by muqiuhan** [comment]
@dotnet-policy-service agree

**Comment by vzarytovskii** [comment]
@DedSec256 had a bunch of changes made to diagnostics context some time ago, maybe he has a couple of thoughts and ideas.

In general - we want to kept few things in mind:

1. We probably want to introduce new style of diagnostics separately instead of changing current one and chose it based on compiler flag and/or current environment (see points below).
2. We need to make sure msbuild/vs/vscode/rider/nvim are rendering new style correctly in their inline diagnostics.
3. We need to make sure that current use environment can correctly render those (for example, terminal supports Unicode, colouring, if we decide to enable it, etc).


In general we do want to make diagnostics prettier and add more context to them, but I think we want a more systematic approach.

**Comment by vzarytovskii** [comment]
Wanted to expand of one of the points - a separate diagnostic style - it will certainly help to not change 100s of tests which match directly on the output, and will allow us to test things separately and make sure we didn't accidentally break anything in the process.

**Comment by psfinaki** [comment]
@muqiuhan thanks for your first PR - welcome to the F# world 😎 

We'll indeed need to discuss this one in a broader context since diags are a very complex topic. Meanwhile, if you are interested in it, feel free to address any of [these problems](https://github.com/dotnet/fsharp/issues?q=is%3Aissue+is%3Aopen+label%3AArea-Diagnostics+) we've already agreed on, some of them are great to get hands dirty here and we'll be happy to guide you and help you with them :)

**Comment by muqiuhan** [comment]
> Wanted to expand of one of the points - a separate diagnostic style - it will certainly help to not change 100s of tests which match directly on the output, and will allow us to test things separately and make sure we didn't accidentally break anything in the process.

Like wrapping the output? I once wrote a little tool that reads the output of the F# compiler and reprocesses it and prints it out:
image

**Comment by muqiuhan** [comment]
> feel free to address any of [these problems](https://github.com/dotnet/fsharp/issues?q=is%3Aissue+is%3Aopen+label%3AArea-Diagnostics+)

Thanks for the tip! I will try to participate in solving these problems ^_^

**Comment by vzarytovskii** [comment]
> > Wanted to expand of one of the points - a separate diagnostic style - it will certainly help to not change 100s of tests which match directly on the output, and will allow us to test things separately and make sure we didn't accidentally break anything in the process.
> 
> Like wrapping the output? I once wrote a little tool that reads the output of the F# compiler and reprocesses it and prints it out: image

No, not wrapping necessary, but as you saw, in `FormatDiagnosticLocation` function, we have a few separate diagnostic styles, in this PR you have changed `DiagnosticStyle.Default`'s output, I suggest we can add a new one, let's say `DiagnosticStyle.Rich` where we can make a new output style, with your changes there.

And we can set it to `Rich` (via `tcConfig.diagnosticStyle`) when certain compiler flag is passed (for example introduce a new flag - `--richdiagnostics+|-`), so we can leave current diagnostics unchanged for time being (so we're sure no tooling is broken accidentally), and test the new format (by just passing the flag), to see any problematic corner cases.

Later on we can swap the default to `Rich` with some fallbacks, like if terminal doesn't support unicode, or is too narrow, we fallback to `DiagnosticStyle.Default`.

Sorry for a bit hectic description, I hope it makes sense :)

**Comment by vzarytovskii** [comment]
If you would like to, I can make this change in your branch (maybe later this week), to demonstrate what I mean.

**Comment by muqiuhan** [comment]
> I suggest we can add a new one, let's say `DiagnosticStyle.Rich` where we can make a new output style, with your changes there.

I understand what you mean, you expressed it very clearly, it's a great idea!

File: src/Compiler/Driver/CompilerDiagnostics.fs [hunk 1]
```diff
@@ -2073,7 +2074,7 @@ let FormatDiagnosticLocation (tcConfig: TcConfig) m : FormattedDiagnosticLocatio
             | DiagnosticStyle.Default ->
                 let file = file.Replace('/', Path.DirectorySeparatorChar)
                 let m = withStart (mkPos m.StartLine (m.StartColumn + 1)) m
-                (sprintf "%s(%d,%d): " file m.StartLine m.StartColumn), m, file
+                (sprintf "◦→ %s (%d,%d)" file m.StartLine m.StartColumn), m, file
```
Comment (baronfel) [inline src/Compiler/Driver/CompilerDiagnostics.fs:2077]: Some context - one of the reason the compiler emits this `<file>(<range>): <message>` format is that this is the [MSBuild Canonical Diagnostic format](https://learn.microsoft.com/en-us/visualstudio/msbuild/msbuild-diagnostic-format-for-tasks?view=vs-2022) - emitting this format helps many tools that operate in MSBuild or use MSBuild (like VS/VSCode/Rider) seamlessly be able to report diagnostics without any additional integration work.

**Comment by vzarytovskii** [comment]
> > I suggest we can add a new one, let's say `DiagnosticStyle.Rich` where we can make a new output style, with your changes there.
> 
> I understand what you mean, you expressed it very clearly, it's a great idea!
> 
> like this:
> 
> ```fsharp
> /// Represents the style being used to format errors
> [<RequireQualifiedAccess>]
> type DiagnosticStyle =
>     | Default
>     | Rich
> ...
> ```
> 
> Then add it elsewhere:
> 
> ```fsharp
> match tcConfig.diagnosticStyle with
> | DiagnosticStyle.Emacs -> ...
> | DiagnosticStyle.Default -> ...
> | DiagnosticStyle.Rich -> ...
> ...
> ```

Yep, precisely. And then we can add a new compiler flag, which will be selecting `DiagnosticStyle.Rich` as default one based on flag.

File: src/Compiler/Driver/CompilerDiagnostics.fs (hunk 1 above)
Comment (edgarfgp) [reply src/Compiler/Driver/CompilerDiagnostics.fs:2077]: Yeah I think we need a more `Rich` diagnostic format the current one is not really friendly and I think we can do better.

**Comment by T-Gro** [comment]
@muqiuhan : I have marked the PR as a draft to avoid accidental merging.

Once this is done as a new DiagnosticsStyle (per the suggestion from Vlad above), this will be ready for merge.

**Comment by vzarytovskii** [comment]
I intend to add a flag tomorrow so we can maybe start testing it sooner? It should take me a couple of hours. I will add a new format for the diagnostic and make tests green (won't be adding any tests now). Let me know if anyone has any objections.

**Comment by vzarytovskii** [comment]
Ok, I've added new flag (`--richerrors`), and pushed changes (sorry, couldn't wait, that's legit one of the first things I wanted to do once I got on the team, and could never start):

image

I propose we push the change to the main, so we can start playing with format and how we output multiple errors, work on some additional things in parallel, etc. I personally think, that "rust-like" format is the ideal one to start with:

```
error FSXXXX:
1 | let z = x = yyyy
                ^^^^
> Error description...

--------

warning FSXXX:
...
...
```
We can then try playing with format, characters we use, multi-line errors, etc.

WDYT?

cc @baronfel @T-Gro @dsyme

**Comment by baronfel** [comment]
(me and @rainersigwald right now)

We're going to be watching what y'all do from the perspective of "what information would a compiler need to give to MSBuild in order for _MSBuild_ to be able to render a message as nicely as this".

**Comment by vzarytovskii** [comment]
>  [  ](https://private-user-images.githubusercontent.com/573979/361489463-ba1b46ba-594a-49e8-8d62-068181c3ba03.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ2ODY3NjYsIm5iZiI6MTcyNDY4NjQ2NiwicGF0aCI6Ii81NzM5NzkvMzYxNDg5NDYzLWJhMWI0NmJhLTU5NGEtNDllOC04ZDYyLTA2ODE4MWMzYmEwMy5naWY_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODI2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgyNlQxNTM0MjZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wMDFlMTdhMzBkZWZkNGQ2ZDAzZDhmMTcxMGVhMzhiMTUyYzNlNTA0MjEyYmQxODcwNDQzYTc4YmVhM2ZiNTJmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.-_M0shMLePBSLf8SdJaTNV6kWbJsPG2BSLHrrfesXCo) [ ](https://private-user-images.githubusercontent.com/573979/361489463-ba1b46ba-594a-49e8-8d62-068181c3ba03.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ2ODY3NjYsIm5iZiI6MTcyNDY4NjQ2NiwicGF0aCI6Ii81NzM5NzkvMzYxNDg5NDYzLWJhMWI0NmJhLTU5NGEtNDllOC04ZDYyLTA2ODE4MWMzYmEwMy5naWY_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODI2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgyNlQxNTM0MjZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wMDFlMTdhMzBkZWZkNGQ2ZDAzZDhmMTcxMGVhMzhiMTUyYzNlNTA0MjEyYmQxODcwNDQzYTc4YmVhM2ZiNTJmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.-_M0shMLePBSLf8SdJaTNV6kWbJsPG2BSLHrrfesXCo)
> 
> (me and @rainersigwald right now)
> 
> We're going to be watching what y'all do from the perspective of "what information would a compiler need to give to MSBuild in order for _MSBuild_ to be able to render a message as nicely as this".

Currently, I'm going to test it with FSI, thought it's applicable for msbuild as well.

**Comment by edgarfgp** [comment]
Great job @muqiuhan and @vzarytovskii. I’m looking forward to see what is next here.

**Review by vzarytovskii — APPROVED** [review]

**Review by edgarfgp — APPROVED** [review]

**Comment by vzarytovskii** [comment]
Happy to merge it if everyone agrees, to start working on some format adjustments

File: src/Compiler/Driver/CompilerDiagnostics.fs [hunk 2]
```diff
@@ -2114,6 +2115,10 @@ let FormatDiagnosticLocation (tcConfig: TcConfig) m : FormattedDiagnosticLocatio
                     sprintf "%s(%d,%d,%d,%d): " file m.StartLine m.StartColumn m.EndLine m.EndColumn, m, file
                 else
                     "", m, file
+            | DiagnosticStyle.Rich ->
+                let file = file.Replace('/', Path.DirectorySeparatorChar)
+                let m = withStart (mkPos m.StartLine (m.StartColumn + 1)) m
+                (sprintf "◦→ %s (%d,%d)" file m.StartLine m.StartColumn), m, file
```
Comment (psfinaki) [inline src/Compiler/Driver/CompilerDiagnostics.fs:2121]: I am not an expert in any encoding stuff, random question: can these symbols accidentally produce some display issues anywhere? Why not just use good old `->`? :)

Comment (vzarytovskii) [reply src/Compiler/Driver/CompilerDiagnostics.fs:2121]: Probably they do. That's why I'd like to merge it as is, and then have couple of PRs with different formats to see which one works/easier to consume

File: src/Compiler/Driver/CompilerDiagnostics.fs
Comment (psfinaki) [inline src/Compiler/Driver/CompilerDiagnostics.fs:1]: Would be good to have some tests on how this plays out, would be even more helpful for followups with ever-improving formatting.

File: src/Compiler/Driver/CompilerDiagnostics.fs (hunk 2 above)
Comment (vzarytovskii) [reply src/Compiler/Driver/CompilerDiagnostics.fs:2121]: Reasoning is that it's strictly opt-in with a hidden flag, not known to anyone

**Review by psfinaki — APPROVED** [review]
These bubbles looks weird to me really - but it's a good move anyway.

**Comment by vzarytovskii** [comment]
> These bubbles looks weird to me really - but it's a good move anyway.

Yea, it's just "make it work" PR

File: src/Compiler/Driver/CompilerDiagnostics.fs
Comment (psfinaki) [reply src/Compiler/Driver/CompilerDiagnostics.fs:1]: Alright then, makes sense.

File: src/Compiler/Driver/CompilerDiagnostics.fs (hunk 2 above)
Comment (psfinaki) [reply src/Compiler/Driver/CompilerDiagnostics.fs:2121]: Alright, I was just curious why this was even picked in the first place.

**Comment by psfinaki** [comment]
Leaving the merge on you here, in case you want anyone else to take a look.
